apps/studio/src/features/editing-experience/components/form-builder/renderers/controls/JsonFormsDgsDatasetIdControl.tsx

```diff
@@ -0,0 +1,311 @@
+    mode: "onChange",
+    schema: z.object({
+      datasetId: z
+        .string()
+        .min(1, "Dataset URL is required")
+        .refine((value) => getDgsIdFromString({ string: value }) !== null, {
+          message:
+            "This doesn't look like a valid link from data.gov.sg. Check that you have the correct link and try again.",
+        }),
+    }),
+    reValidateMode: "onChange",
+  })
+
+  // Update inputValue when initialValue changes (when modal opens)
+  useEffect(() => {
+    setInputValue(initialValueUrl)
+  }, [initialValueUrl])
+
+  // Handle dataset validation
+  useEffect(() => {
+    if (isValidatingDataset || !datasetId) return
+
+    if (isValidDataset) {
+      clearErrors("datasetId")
+      return
+    }
+
+    setError("datasetId", {
+      type: "manual",
+      message: format
+        ? "You can only link CSV datasets. Please check the dataset ID and try again."
+        : "This doesn’t look like a valid link from data.gov.sg. Check that you have the correct link and try again.",
+    })
+  }, [
+    datasetId,
+    isValidDataset,
+    format,
+    isValidatingDataset,
+    setError,
+    clearErrors,
+  ])
+
+  const onSubmit = handleSubmit(() => {
+    const extractedId = getDgsIdFromString({ string: debouncedInputValue })
+    if (extractedId) {
+      onClose()
+      onSave(extractedId) // Save only the ID, not the full URL
+    }
+  })
+
+  const handleClose = () => {
+    onClose()
+    setInputValue(initialValueUrl)
+  }
+
+  const FeedbackMessage = () => {
+    if (errors.datasetId) {
+      return <FormErrorMessage>{errors.datasetId.message}</FormErrorMessage>
+    }
+    if (isLoading) {
+      return (
+        <Text fontSize="sm" color="base.content.medium" mt="0.5rem">
+          Validating dataset...
+        </Text>
+      )
+    }
+    if (isValidDataset) {
+      return (
+        <Text fontSize="sm" color="green.600" mt="0.5rem">
+          ✓ Valid CSV dataset
+        </Text>
+      )
+    }
+  }
+
+  return (
+    <Modal isOpen={isOpen} onClose={handleClose}>
+      <ModalOverlay />
+      <ModalContent>
+        <form onSubmit={onSubmit}>
+          <ModalHeader mr="3.5rem">Link a dataset</ModalHeader>
+          <ModalCloseButton size="lg" />
+
+          <ModalBody>
+            <FormControl mt="1rem" isRequired isInvalid={!!errors.datasetId}>
+              <FormLabel description="Navigate to your data on Data.gov.sg and copy the dataset’s link">
+                Link to your CSV dataset
+              </FormLabel>
+
+              <Input
+                fontFamily="monospace"
+                placeholder="Paste dataset URL here"
+                value={inputValue}
+                onChange={(e) => {
+                  setInputValue(e.target.value)
+                  void register("datasetId").onChange(e)
+                }}
+                onBlur={register("datasetId").onBlur}
+                name={register("datasetId").name}
+                ref={register("datasetId").ref}
```

Comment:
issue: we should spread the `register("datasetId")` here, then override just the `onChange`